### ISSUE TITLE
Docs: Playground not emitting some type-aware lint errors

### ISSUE THREAD

**Issue by Josh-Cena** [typescript-eslint/typescript-eslint#5142]
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

It seems some type-aware rules are not emitting any errors in the playground.

Setting `"@typescript-eslint/no-base-to-string": "error"`, and writing `"" + {}` doesn't trigger errors.

[Playground](https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=ETAEGpQbwXwbgFBA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oCMBDZRLXxdk+KkwDm6KImjQO0SODABfECqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

Floating promises don't trigger errors.

[Playground](https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=IYZwngdgxgBAZgV2gFwJYHsL3egFAShgG8YAnAU2QVKwEYBuGAXwChEUMsAjYUg4ljGx589FqyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

Type-unaware rules do work.

Maybe I'm doing something wrong? It's at least confusing that configurations don't immediately trigger errors.

### Affected URL(s)

https://typescript-eslint.io/play

**Comment by armano2** [comment]
i got same issue while trying to reproduce #4999, it seem that we have issue somwhere

**Comment by armano2** [comment]
after some digging and aligning types from eslint with those used in editor, i found out an issue, 

![image](https://user-images.githubusercontent.com/625469/172702742-52a7fcd9-5fac-4b0c-8d99-1bb3ea87a64d.png)

-------------

lib resolution is a little messy and its not workign correctly as it should (it never did),

if you add to tsconfig it should start working
```
    "lib": ["esnext", "es2021"]
```
